In [1]:
import numpy as np
import pandas as pd
from sklearn.utils import class_weight

In [3]:
# df = pd.read_excel('one_hot_encode2004.xlsx')
df = pd.read_excel('cleanData_Label.xlsx')
# df1.drop(['other_pro_lang'], axis=1, inplace=True)
# label = df1['Class Label']
# df1.drop(['Class Label'], axis=1, inplace=True)
# df = pd.concat([df1, label], axis=1)
df

age(y)  career  exp(y)  language  web_pro_lang  UI_Libs  front_dev  \
0         2       2       1         1             1        2          1   
1         2       2       2         1             1        1          1   
2         2       4       2         1             1        1          1   
3         2       2       2         2             2        1          1   
4         2       2       2         1             2        2          1   
..      ...     ...     ...       ...           ...      ...        ...   
292       2       2       2         1             1        3          2   
293       2       2       1         1             6        2          2   
294       2       2       2         1             1        2          2   
295       2       2       1         1             4        1          2   
296       2       2       1         1             4        1          2   

     learning_src  duration(m)  factor_market_needs  factor_compensation  \
0               1            2                    2                    2   
1               2            2                    2                    2   
2               3            2                    3                    3   
3               4            2                    3                    3   
4               5            2                    3                    3   
..            ...          ...                  ...                  ...   
292             8            1                    3                    3   
293             2            1                    2                    1   
294            11            2                    3                    3   
295             1            1                    1                    1   
296             1            1                    1                    1   

     website_function_present  website_content  website_graphics  \
0                           3                1                 2   
1                           3                1                 2   
2                           3                1                 2   
3                           3                2                 2   
4                           3                2                 2   
..                        ...              ...               ...   
292                         3                2                 3   
293                         2                3                 1   
294                         3                1                 2   
295                         3                1                 2   
296                         3                1                 2   

     website_functions Class Label  
0                    1       React  
1                    2      Vue.js  
2                    3       React  
3                    2      Vue.js  
4                    3     Angular  
..                 ...         ...  
292                  2       React  
293                  1      Vue.js  
294                  2      jQuery  
295                  2     Angular  
296                  2     Angular  

[297 rows x 16 columns]

In [4]:
# x = df.iloc[:, 0:15]
# y = df.iloc[:, -1]

In [5]:
array = df.values

In [6]:
import numpy as np
X = array[:,0:15]
Y = array[:,15]

In [7]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

In [8]:
# Feature extraction
test = SelectKBest(score_func=chi2, k=4)
fit = test.fit(X, Y)

# Summarize scores
np.set_printoptions(precision=3)
print(fit.scores_)

features = fit.transform(X)
# Summarize selected features
# print(features[0:15,:])
print(df.columns)

[1.609 0.561 1.162 1.416 2.754 3.245 0.145 5.043 0.809 2.684 2.882 0.424
 0.99  2.28  0.177]
Index(['age(y)', 'career', 'exp(y)', 'language', 'web_pro_lang', 'UI_Libs',
       'front_dev', 'learning_src', 'duration(m)', 'factor_market_needs',
       'factor_compensation', 'website_function_present', 'website_content',
       'website_graphics', 'website_functions', 'Class Label'],
      dtype='object')


'age(y)', 1.609 <br>
'career', 0.561 x <br>
'exp(y)', 1.162  <br>
'language', 1.416  <br>
'web_pro_lang', 2.754 <br>
'UI_Libs',3.245  <br>
'front_dev', 0.145 x <br>
'learning_src', 5.043  <br>
'duration(m)', 0.809 x <br>
'factor_market_needs',2.684  <br>
'factor_compensation', 2.882  <br>
'website_function_present', 0.424 <br>
'website_content',0.424 x <br>
'website_graphics', 2.28  <br>
'website_functions', 0.177 x 

In [9]:
df.drop(['career'], axis=1, inplace=True)
df.drop(['front_dev'], axis=1, inplace=True)
df.drop(['duration(m)'], axis=1, inplace=True)
df.drop(['website_content'], axis=1, inplace=True)
df.drop(['website_functions'], axis=1, inplace=True)

In [10]:
df

age(y)  exp(y)  language  web_pro_lang  UI_Libs  learning_src  \
0         2       1         1             1        2             1   
1         2       2         1             1        1             2   
2         2       2         1             1        1             3   
3         2       2         2             2        1             4   
4         2       2         1             2        2             5   
..      ...     ...       ...           ...      ...           ...   
292       2       2         1             1        3             8   
293       2       1         1             6        2             2   
294       2       2         1             1        2            11   
295       2       1         1             4        1             1   
296       2       1         1             4        1             1   

     factor_market_needs  factor_compensation  website_function_present  \
0                      2                    2                         3   
1                      2                    2                         3   
2                      3                    3                         3   
3                      3                    3                         3   
4                      3                    3                         3   
..                   ...                  ...                       ...   
292                    3                    3                         3   
293                    2                    1                         2   
294                    3                    3                         3   
295                    1                    1                         3   
296                    1                    1                         3   

     website_graphics Class Label  
0                   2       React  
1                   2      Vue.js  
2                   2       React  
3                   2      Vue.js  
4                   2     Angular  
..                ...         ...  
292                 3       React  
293                 1      Vue.js  
294                 2      jQuery  
295                 2     Angular  
296                 2     Angular  

[297 rows x 11 columns]

In [11]:
df.to_excel('dataFeature.xlsx',index=False,encoding='utf-8-sig')

In [12]:
x = df.iloc[:, 0:10]
y = df.iloc[:, -1]

### Decision Tree

In [13]:
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier(criterion="entropy", class_weight='balanced')
clf = clf.fit(x, y)

In [14]:
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report 

y_pred = cross_val_predict(clf, x, y, cv=10)

scores = cross_val_score(clf, x, y, cv=10)
print("cross-validated scores : ", scores)

avg_score = np.mean(scores)
print("\n cross-validated avg score : ", avg_score)

confusion_matrix = pd.crosstab(y, y_pred, rownames=['Actual'], colnames=['Predicted'], margins=True)
print("\n", confusion_matrix)

print("\n", classification_report(y, y_pred))

cross-validated scores :  [0.4   0.4   0.267 0.5   0.367 0.3   0.367 0.448 0.31  0.448]

 cross-validated avg score :  0.3806896551724138

 Predicted  Angular  React  Vue.js  jQuery  All
Actual                                        
Angular         23     18      11       5   57
React           20     62      30      21  133
Vue.js          12     23      13      17   65
jQuery           5     14      14       9   42
All             60    117      68      52  297

               precision    recall  f1-score   support

     Angular       0.38      0.40      0.39        57
       React       0.53      0.47      0.50       133
      Vue.js       0.19      0.20      0.20        65
      jQuery       0.17      0.21      0.19        42

    accuracy                           0.36       297
   macro avg       0.32      0.32      0.32       297
weighted avg       0.38      0.36      0.37       297



In [15]:
clf = DecisionTreeClassifier(criterion="entropy", max_depth=5)
clf = clf.fit(x, y)
y_pred = cross_val_predict(clf, x, y, cv=10)

scores = cross_val_score(clf, x, y, cv=10)
print("cross-validated scores : ", scores)

avg_score = np.mean(scores)
print("\n cross-validated avg score : ", avg_score)

confusion_matrix = pd.crosstab(y, y_pred, rownames=['Actual'], colnames=['Predicted'], margins=True)
print("\n", confusion_matrix)

print("\n", classification_report(y, y_pred))

cross-validated scores :  [0.467 0.467 0.433 0.533 0.533 0.333 0.533 0.517 0.345 0.621]

 cross-validated avg score :  0.47827586206896555

 Predicted  Angular  React  Vue.js  jQuery  All
Actual                                        
Angular         22     26       8       1   57
React           12    103      12       6  133
Vue.js           7     42      13       3   65
jQuery           4     25      10       3   42
All             45    196      43      13  297

               precision    recall  f1-score   support

     Angular       0.49      0.39      0.43        57
       React       0.53      0.77      0.63       133
      Vue.js       0.30      0.20      0.24        65
      jQuery       0.23      0.07      0.11        42

    accuracy                           0.47       297
   macro avg       0.39      0.36      0.35       297
weighted avg       0.43      0.47      0.43       297



In [16]:

clf.predict([[2,1,1,1,1,1,1,1,1,1]])

array(['Angular'], dtype=object)

In [17]:
# # from sklearn.externals.six import StringIO  
# from six import StringIO
# from IPython.display import Image  
# from sklearn.tree import export_graphviz
# import pydotplus

# df_feature_names = x.columns
# df_target_names = [str(s) for s in y.unique()]

# dot_data = StringIO()
# export_graphviz(clf, out_file=dot_data, feature_names=df_feature_names, class_names=df_target_names)
# graph = pydotplus.graph_from_dot_data(dot_data.getvalue())  
# Image(graph.create_png())

In [18]:
# clf = DecisionTreeClassifier(criterion="entropy", max_depth=5)
# clf = clf.fit(x, y)
# y_pred = cross_val_predict(clf, x, y, cv=10)

# scores = cross_val_score(clf, x, y, cv=10)
# print("cross-validated scores : ", scores)

# avg_score = np.mean(scores)
# print("\n cross-validated avg score : ", avg_score)

# confusion_matrix = pd.crosstab(y, y_pred, rownames=['Actual'], colnames=['Predicted'], margins=True)
# print("\n", confusion_matrix)

# print("\n", classification_report(y, y_pred))

In [19]:
# # from sklearn.externals.six import StringIO  
# from six import StringIO
# from IPython.display import Image  
# from sklearn.tree import export_graphviz
# import pydotplus

# df_feature_names = x.columns
# df_target_names = [str(s) for s in y.unique()]

# dot_data = StringIO()
# export_graphviz(clf, out_file=dot_data, feature_names=df_feature_names, class_names=df_target_names)
# graph = pydotplus.graph_from_dot_data(dot_data.getvalue())  
# Image(graph.create_png())

### SVM

In [20]:
from sklearn import svm

clf_svm = svm.SVC(class_weight='balanced')
clf_svm.fit(x,y)

SVC(class_weight='balanced')

In [21]:
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report 

y_pred = cross_val_predict(clf_svm, x, y, cv=10)

scores = cross_val_score(clf_svm, x, y, cv=10)
print("cross-validated scores : ", scores)

avg_score = np.mean(scores)
print("\ncross-validated avg score : ", avg_score)

confusion_matrix = pd.crosstab(y, y_pred, rownames=['Actual'], colnames=['Predicted'], margins=True)
print("\n", confusion_matrix)

print("\n", classification_report(y, y_pred))

cross-validated scores :  [0.467 0.2   0.367 0.433 0.433 0.2   0.3   0.379 0.31  0.31 ]

cross-validated avg score :  0.33999999999999997

 Predicted  Angular  React  Vue.js  jQuery  All
Actual                                        
Angular         12     16      20       9   57
React           19     52      41      21  133
Vue.js           8     16      26      15   65
jQuery           5     12      14      11   42
All             44     96     101      56  297

               precision    recall  f1-score   support

     Angular       0.27      0.21      0.24        57
       React       0.54      0.39      0.45       133
      Vue.js       0.26      0.40      0.31        65
      jQuery       0.20      0.26      0.22        42

    accuracy                           0.34       297
   macro avg       0.32      0.32      0.31       297
weighted avg       0.38      0.34      0.35       297



### KNN

In [22]:
from sklearn.neighbors import KNeighborsClassifier

clf_knn = KNeighborsClassifier(n_neighbors=5, metric='euclidean', weights='uniform')
clf_knn.fit(x,y)

KNeighborsClassifier(metric='euclidean')

In [23]:
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report 

y_pred = cross_val_predict(clf_knn, x, y, cv=10)

scores = cross_val_score(clf_knn, x, y, cv=10)
print("cross-validated scores : ", scores)

avg_score = np.mean(scores)
print("\ncross-validated avg score : ", avg_score)

confusion_matrix = pd.crosstab(y, y_pred, rownames=['Actual'], colnames=['Predicted'], margins=True)
print("\n", confusion_matrix)

print("\n", classification_report(y, y_pred))

cross-validated scores :  [0.467 0.467 0.5   0.333 0.5   0.5   0.433 0.31  0.276 0.586]

cross-validated avg score :  0.4372413793103448

 Predicted  Angular  React  Vue.js  jQuery  All
Actual                                        
Angular         20     27       8       2   57
React           23     92      17       1  133
Vue.js           4     43      15       3   65
jQuery           6     29       4       3   42
All             53    191      44       9  297

               precision    recall  f1-score   support

     Angular       0.38      0.35      0.36        57
       React       0.48      0.69      0.57       133
      Vue.js       0.34      0.23      0.28        65
      jQuery       0.33      0.07      0.12        42

    accuracy                           0.44       297
   macro avg       0.38      0.34      0.33       297
weighted avg       0.41      0.44      0.40       297



### Naive Bayes

In [24]:
from sklearn.naive_bayes import GaussianNB
from sklearn.utils.class_weight import compute_sample_weight

weight_balanced = compute_sample_weight(class_weight='balanced', y=y)

clf_naive = GaussianNB()
clf_naive.fit(x,y, sample_weight=weight_balanced)

GaussianNB()

In [25]:
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report 

y_pred = cross_val_predict(clf_naive, x, y, cv=10)

scores = cross_val_score(clf_naive, x, y, cv=10)
print("cross-validated scores : ", scores)

avg_score = np.mean(scores)
print("\ncross-validated avg score : ", avg_score)

confusion_matrix = pd.crosstab(y, y_pred, rownames=['Actual'], colnames=['Predicted'], margins=True)
print("\n", confusion_matrix)

print("\n", classification_report(y, y_pred))

cross-validated scores :  [0.367 0.433 0.433 0.433 0.4   0.3   0.4   0.448 0.448 0.448]

cross-validated avg score :  0.41114942528735626

 Predicted  Angular  React  Vue.js  jQuery  All
Actual                                        
Angular          4     21      30       2   57
React            3     86      31      13  133
Vue.js           3     29      27       6   65
jQuery           5     15      17       5   42
All             15    151     105      26  297

               precision    recall  f1-score   support

     Angular       0.27      0.07      0.11        57
       React       0.57      0.65      0.61       133
      Vue.js       0.26      0.42      0.32        65
      jQuery       0.19      0.12      0.15        42

    accuracy                           0.41       297
   macro avg       0.32      0.31      0.30       297
weighted avg       0.39      0.41      0.38       297



------

In [20]:
import pickle

In [21]:
pkl_filename = "front_end_model.pkl"

with open(pkl_filename, 'wb') as file:
  pickle.dump(clf, file)